### PR TITLE
client only build

### PR DESCRIPTION
**Pull request by cmickeyb**
enables a client-only build that does not require SGX to be installed. this build would be used by most clients who are building or interacting with contracts. biggest change is a fairly substantial overhaul of the cmake files in common to remove unnecessary dependencies on sgx sdk and sgx ssl.

the current docker tests do not exercise the client. to test you'll need set of running services (probably on a different host) then use make test in wawaka/contracts.

### CODE REVIEW

File: common/CMakeLists.txt [hunk 1]
```diff
@@ -13,99 +13,120 @@
...
+  TARGET_INCLUDE_DIRECTORIES(${T_COMMON_LIB_NAME} PRIVATE ${SGX_SDK}/include)
+  TARGET_INCLUDE_DIRECTORIES(${T_COMMON_LIB_NAME} PRIVATE ${SGX_SSL}/include)
+  TARGET_INCLUDE_DIRECTORIES(${T_COMMON_LIB_NAME} PRIVATE ${SGX_SDK}/include/tlibc)
+  TARGET_INCLUDE_DIRECTORIES(${T_COMMON_LIB_NAME} PRIVATE ${SGX_SDK}/include/libcxx)
 
+  TARGET_COMPILE_OPTIONS(${T_COMMON_LIB_NAME} PRIVATE ${COMMON_CXX_FLAGS})
+  TARGET_COMPILE_OPTIONS(${T_COMMON_LIB_NAME} PRIVATE -nostdinc)
+  TARGET_COMPILE_OPTIONS(${T_COMMON_LIB_NAME} PRIVATE -nostdinc++)
+  TARGET_COMPILE_OPTIONS(${T_COMMON_LIB_NAME} PRIVATE -fno-builtin-printf)
 endif()
```
Comment (marcelamelara) [inline common/CMakeLists.txt:110]: Yes, this is a total nit.

```suggestion
ENDIF()
```

File: common/crypto/CMakeLists.txt [hunk 2]
```diff
@@ -14,89 +14,101 @@
...
-    SET(U_CRYPTO_LIB_NAME updo-crypto)
-ENDIF()
-IF("${T_CRYPTO_LIB_NAME} " STREQUAL " ")
-    SET(T_CRYPTO_LIB_NAME tpdo-crypto)
-ENDIF()
-
-FILE(GLOB PROJECT_HEADERS *.h verify_ias_report/*.h)
-FILE(GLOB PROJECT_SOURCES *.cpp verify_ias_report/*.cpp)
+# NOTE: this file is included by the common CMakeLists.txt;
+# it should not be evaluated independently
```
Comment (marcelamelara) [inline common/crypto/CMakeLists.txt:18]: Wondering if we could detect this case somehow, and throw a fatal error.

File: common/error.cpp [hunk 3]
```diff
@@ -186,3 +188,5 @@ namespace pdo {
         } // ThrowSgxError
     }
 } // namespace pdo
+
+#endif // _CLIENT_ONLY_
```
Comment (marcelamelara) [inline common/error.cpp:192]: Shouldn't we just omit `error.cpp` from the client build in cmake? I'd imagine stdlib and stdio are imported into the client through other files.

File: common/interpreter/wawaka_wasm/CMakeLists.txt [hunk 4]
```diff
@@ -40,137 +40,127 @@ SET(WASM_MEM_CONFIG "$ENV{WASM_MEM_CONFIG}")
...
-  endif ()
-endif ()
+    SET (WAMR_BUILD_TARGET "X86_32")
+  ENDIF ()
+ENDIF ()
 
 if (NOT CMAKE_BUILD_TYPE)
-  set(CMAKE_BUILD_TYPE Release)
+  SET(CMAKE_BUILD_TYPE Release)
 endif ()
```
Comment (marcelamelara) [inline common/interpreter/wawaka_wasm/CMakeLists.txt:59]: A few more nitpicky changes for consistency.

```suggestion
ENDIF()
```

File: common/interpreter/wawaka_wasm/CMakeLists.txt [hunk 5]
```diff
@@ -40,137 +40,127 @@ SET(WASM_MEM_CONFIG "$ENV{WASM_MEM_CONFIG}")
...
+  ELSE ()
     # Build as X86_32 by default in 32-bit platform
-    set (WAMR_BUILD_TARGET "X86_32")
-  endif ()
-endif ()
+    SET (WAMR_BUILD_TARGET "X86_32")
+  ENDIF ()
+ENDIF ()
 
 if (NOT CMAKE_BUILD_TYPE)
```
Comment (marcelamelara) [inline common/interpreter/wawaka_wasm/CMakeLists.txt:57]: More nits.

```suggestion
IF (NOT CMAKE_BUILD_TYPE)
```

File: docs/client_install.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+## <a name="environment">Setup Build Environment</a>
+
+On a minimal Ubuntu system, the following packages are required. Other
+distributions will require similar packages.
+
+```bash
+sudo apt install -y cmake curl git pkg-config unzip xxd libssl-dev build-essential
+sudo apt install -y swig python3 python3-dev python3-venv virtualenv
+sudo apt install -y liblmdb-dev libprotobuf-dev libsecp256k1-dev protobuf-compiler libncurses5-dev
```
Comment (marcelamelara) [inline docs/client_install.md:28]: The protobuf dependencies are for sawtooth, right? Does the client need these?

File: docs/client_install.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+```
+
+<!--
+    virtualenv will cause python3 and python3-virtualenv to be installed
+-->
+
+## Build and Install PDO
+
+Assuming you have installed and configured the pre-requisites in the
+default location lcations, the following commands will build and install
```
Comment (marcelamelara) [inline docs/client_install.md:38]:
```suggestion
default location, the following commands will build and install
```

**Review by bvavala — APPROVED** [review]
looks good -- tested in HW mode

File: common/error.cpp (hunk 3 above)
Comment (cmickeyb) [reply common/error.cpp:192]: More concerned about how/where this gets included in other common source.

File: docs/client_install.md (hunk 6 above)
Comment (cmickeyb) [reply docs/client_install.md:28]: We need them for the client if the client is for a sawtooth ledger. So until we drop sawtooth, they are necessary.

**Review by marcelamelara — APPROVED** [review]

**Pull request merged**